rpc_util.go

```diff
@@ -233,6 +234,29 @@ func (o TrailerCallOption) after(c *callInfo, attempt *csAttempt) {
 	*o.TrailerAddr = attempt.s.Trailer()
 }
 
+// Authority returns a CallOption that mofifies the :authority header for a RPC.
+func Authority(modifier func(ctx context.Context, authority string) string) CallOption {
```

Comment:
I'm not fully confident about the signature of this `modifier` function (this is a user-facing API), but at least, I think it's worth making it have `context.Context` as its argument since we can use values inside the Context like the metadata as I described in the description of this PR.